sdk/core/azure-core/azure/core/rest/_http_response_impl_async.py

```diff
@@ -0,0 +1,110 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+from typing import AsyncIterator
+from ._rest_py3 import AsyncHttpResponse as _AsyncHttpResponse
+from ._http_response_impl import _HttpResponseBaseImpl
+from ._helpers import parse_lines_from_text
+
+class AsyncHttpResponseImpl(_HttpResponseBaseImpl, _AsyncHttpResponse):
+    """AsyncHttpResponseImpl built on top of our HttpResponse protocol class.
+
+    Helper impl for creating our transport responses
+    """
+
+    async def read(self) -> bytes:
+        """Read the response's bytes into memory.
+
+        :return: The response's bytes
+        :rtype: bytes
+        """
+        if self._content is None:
+            parts = []
+            async for part in self.iter_bytes():
+                parts.append(part)
+            self._content = b"".join(parts)
+        self._is_stream_consumed = True
+        return self._content
+
+    async def iter_text(self) -> AsyncIterator[str]:
+        """Asynchronously iterates over the text in the response.
+
+        :return: An async iterator of string. Each string chunk will be a text from the response
+        :rtype: AsyncIterator[str]
+        """
+        async for byte in self.iter_bytes():  # type: ignore
+            text = byte.decode(self.encoding or "utf-8")
```

Comment:
What gives us the guarantee that a byte chunk can be decoded (e.g. you don't get a partial multi code-point grapheme)? This looks very scary to me...

Reply:
Yeah you're right I should've updated this logic when I updated the encoding flow. I'm going to call this decoder instead basically, where first we either get the encoding from
1. What the user set
2. What we got from the Content Type response header
3. utf-8-sig
That will then be consistent with our behavior for text decoding in the response's `text` method

Reply:
Talked offline, going to remove iter_text and iter_lines for now